### PR TITLE
[FLINK-22055][runtime] Fix RPCEndpoint MainThreadExecutor scheduling command with wrong time unit.

### PR DESCRIPTION
**Pull request by est08zw**
## What is the purpose of the change

fix typo, use the delay in millisecond rather than the raw parameter which may not in TimeUnit.MILLISECONDS precision.

## Brief change log

fix typo, use the delay in millisecond rather than the raw parameter which may not in TimeUnit.MILLISECONDS precision.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Comment by zentol** [comment]
ping @xintongsong

**Review by xintongsong — COMMENTED** [review]
Thanks @est08zw. Nice catch! The fix LGTM.

I think this also reveals missing of tests covering these scheduling methods. Do you wan to add the tests in this PR? I'm also good with merging this PR as is, and add the tests in a separate issue. WDYT?

**Comment by est08zw** [comment]
@xintongsong
I think it is better the tests and this change in one PR for future review more clearly.

**Comment by xintongsong** [comment]
Thanks @est08zw. Then please ping me once you have the tests ready for review.

**Comment by est08zw** [comment]
ping @xintongsong 

For scheduling runnable with some delay, I find it is hard to check that the delay is exactly the right number.
for example, we schedule a runnable with 1 second delay, and after exactly 1 second, the runnable is always not triggered. So I add extra 50ms delay to check whether the runnable is triggered. I am not very sure it is bug or just because it takes some time for the Executor to execute the runnable.

For scheduling runnable with different TimeUnit, I just checked minute/second/millisecond. For microsecond and nanosecond, i am afraid the executor can not be that accurate. For hour and day, it just take too much time to finish the test.

**Review by xintongsong — CHANGES_REQUESTED** [review]
@est08zw, thanks for adding the tests. I have some comments regarding the test cases, please take a look.

To make the comments easier to understand, there's one example how the tests may look like.

```
@Test
    public void testScheduleRunnable() throws Exception {
        final Time expectedDelay1 = Time.seconds(1);
        final Time expectedDelay2 = Time.milliseconds(500);
        final CompletableFuture<Long> actualDelayMsFuture1 = new CompletableFuture<>();
        final CompletableFuture<Long> actualDelayMsFuture2 = new CompletableFuture<>();

        final RpcEndpoint endpoint = new BaseEndpoint(rpcService);

        try {
            endpoint.start();

            final long startTime = System.currentTimeMillis();
            endpoint.getMainThreadExecutor()
                    .schedule(
                            () -> {
                                // no need to catch the validation failure
                                // if the validation fail, the future will never complete
                                endpoint.validateRunsInMainThread();
                                actualDelayMsFuture1.complete(
                                        System.currentTimeMillis() - startTime);
                            },
                            expectedDelay1.getSize(),
                            expectedDelay1.getUnit());
            endpoint.getMainThreadExecutor()
                    .schedule(
                            () -> {
                                // no need to catch the validation failure
                                // if the validation fail, the future will never complete
                                endpoint.validateRunsInMainThread();
                                actualDelayMsFuture2.complete(
                                        System.currentTimeMillis() - startTime);
                            },
                            expectedDelay2.getSize(),
                            expectedDelay2.getUnit());

            final long actualDelayMs1 =
                    actualDelayMsFuture1.get(
                            expectedDelay1.getSize() * 2, expectedDelay1.getUnit());
            final long actualDelayMs2 =
                    actualDelayMsFuture2.get(
                            expectedDelay2.getSize() * 2, expectedDelay2.getUnit());

            assertTrue(actualDelayMs1 > expectedDelay1.toMilliseconds() * 0.8);
            assertTrue(actualDelayMs1 < expectedDelay1.toMilliseconds() * 1.8);
            assertTrue(actualDelayMs2 > expectedDelay2.toMilliseconds() * 0.8);
            assertTrue(actualDelayMs2 < expectedDelay2.toMilliseconds() * 1.8);
        } finally {
            RpcUtils.terminateRpcEndpoint(endpoint, TIMEOUT);
        }
    }
```

**Comment by xintongsong** [comment]
@est08zw,
I have opened FLINK-22055 for this issue and attached this PR to the ticket.
Please leave a message on the ticket, so I know your jira id and can assign you to the ticket.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java [hunk 1]
```diff
@@ -262,4 +267,136 @@ protected ExtendedEndpoint(
...
         }
     }
+
+    /**
+     * Tests that the {@link RpcEndpoint} can execute the runnable. See also {@link
+     * org.apache.flink.runtime.rpc.RpcEndpoint#runAsync(Runnable)}
+     */
+    @Test
+    public void testRunAsync() throws InterruptedException, ExecutionException, TimeoutException {
+        RpcEndpoint dummyRpcEndpoint = new RpcEndpoint(rpcService) {};
```
[the remark this answers is not in the document]
Comment (est08zw) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java]: add one parameter constructor for BaseEndpoint and use baseEndpoint for the tests.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java [hunk 2]
```diff
@@ -262,4 +267,136 @@ protected ExtendedEndpoint(
...
+        final OneShotLatch latch = new OneShotLatch();
+        try {
+            dummyRpcEndpoint.start();
+            dummyRpcEndpoint.runAsync(latch::trigger);
```
[the remark this answers is not in the document]
Comment (est08zw) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java]: use RpcEndpoint#validateRunsInMainThread() to check run in the main thread.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java [hunk 3]
```diff
@@ -262,4 +267,136 @@ protected ExtendedEndpoint(
...
+     */
+    @Test
+    public void testScheduleRunAsyncTime()
+            throws InterruptedException, ExecutionException, TimeoutException {
+        RpcEndpoint dummyRpcEndpoint = new RpcEndpoint(rpcService) {};
+        OneShotLatch latch = new OneShotLatch();
+        Time delay = Time.seconds(1);
+        try {
+            dummyRpcEndpoint.start();
+            dummyRpcEndpoint.scheduleRunAsync(latch::trigger, delay);
```
[the remark this answers is not in the document]
Comment (est08zw) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java]: use RpcEndpoint#getMainThreadExecutor() to schedule command indirectly.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java [hunk 4]
```diff
@@ -262,4 +267,136 @@ protected ExtendedEndpoint(
...
+        }
+    }
+
+    /**
+     * Tests that the {@link RpcEndpoint} can execute the runnable with a delay specified in
+     * TimeUnit. See also {@link org.apache.flink.runtime.rpc.RpcEndpoint#scheduleRunAsync(Runnable,
+     * long, TimeUnit)}
+     */
+    @Test
+    public void testScheduleRunAsyncTimeUnit()
```
[the remark this answers is not in the document]
Comment (est08zw) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java]: just test MILLISECONDS and SECONDS. reduce the delay and schedule the 2 commands immediately, do not wait the first command finished to schedule the 2nd command.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java [hunk 5]
```diff
@@ -262,4 +267,136 @@ protected ExtendedEndpoint(
...
+            dummyRpcEndpoint.start();
+
+            dummyRpcEndpoint.scheduleRunAsync(latch::trigger, delayMinutes, TimeUnit.MINUTES);
+            assertFalse(latch.isTriggered());
+            if (addExtraDelay) {
+                TimeUnit.MILLISECONDS.sleep(extraDelayMilliSeconds);
+            }
+            latch.await(delayTime.getSize(), delayTime.getUnit());
+            assertTrue(latch.isTriggered());
+            latch.reset();
```
[the remark this answers is not in the document]
Comment (est08zw) [reply flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java]: use [delay * 0.8, delay * 1.2] range to check the actual delay.

File: flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java [hunk 6]
```diff
@@ -262,4 +269,268 @@ protected ExtendedEndpoint(
...
+            endpoint.start();
+            CompletableFuture<Integer> integerFuture =
+                    endpoint.callAsync(
+                            () -> {
+                                endpoint.validateRunsInMainThread();
+                                TimeUnit.MILLISECONDS.sleep(300);
+                                return expectedInteger;
+                            },
+                            timeout);
+            integerFuture.get(timeout.getSize(), timeout.getUnit());
```
Comment (xintongsong) [inline flink-runtime/src/test/java/org/apache/flink/runtime/rpc/RpcEndpointTest.java]: This should not work.

```
future = endpoint.callAsync(callable, timeout1);
future.get(timeout2.getSize(), timeout2.getUnit);
```
I think this case should test whether `timeout1` in the above example takes effect, while currently the `TimeoutException` comes from `timeout2`. You can verify that by replace `integerFuture.get(timeout.getSize(), timeout.getUnit())` with `integerFuture.get()`. An `ExecutionException` should be thrown.

When the callable execution takes longer than `timeout2`, `integerFuture` will be completed exceptionally. Trying to `get` a `CompletableFuture` that is completed exceptionally will result in the `ExecutionException`. Instead, we should verify that `integerFuture` is completed with a timeout exception.

**Review by xintongsong — APPROVED** [review]
Thanks for addressing the comments, @est08zw.
The changes already look quite good to me. I have only one comment concerning the last test case, plus some minor issues.
I have addressed the comments myself and updated the PR. Please take a look.
If there's no objections, I'll merge this PR once AZP gives green light.

**Pull request closed without merging**